### PR TITLE
Drop Report

### LINKED ISSUE #239: Add drop report to Tofino TNA
Will need to implement counters in aggregate_tna.p4

### PR DESCRIPTION
**Pull request by ghost**
#### What does this PR do?
Fixes #239 
Creates drop report packets that are sent to the AE from the SDN controller in periodic intervals.
The drop report contains three metadata fields - 
- Timestamp
- Table entry identifier (hash of the match keys)
- Drop count
#### How can the reviewer verify this PR?
Run tcpdump on the AE to capture drop report (TNA Lab trial scenario)
#### Questions:
- Have you connected this PR to the issue it resolves? Yes
- Does the documentation need an update? Yes
- Does this add new dependencies? No
- Have you added unit or functional tests for this PR?

### CODE REVIEW

**Comment by Sanjhana-Jayagopal** [comment]
Run Tofino lab trial build

**Review by spisarski — CHANGES_REQUESTED** [review]
Code looks nice and clean. Just one big problem with obtaining the AE IP. This will have to be done via a query to the TpsCoreEgress.setup_telem_rpt_t table in core_tna.p4. Sorry this is a big ask but it is very important that each Telemetry Report obtain its information in the same manner.

P.S. You don't have to deal with IPv6. If you decide not to add that into this PR, please create a task within the backlog.

File: trans_sec/controller/ddos_sdn_controller.py [hunk 1]
```diff
@@ -45,14 +56,57 @@ def start(self, add_di):
...
+        self.create_drop_report()
         self.__main_loop()
 
     def stop(self):
         logger.info('Stopping SDN controller')
         self.running = False
         self.http_server.stop()
 
+    def create_drop_report(self):
+        for controller in self.controllers.values():
```
Comment (spisarski) [inline trans_sec/controller/ddos_sdn_controller.py:68]: You could add a call to the core_controller here (that calls a new method on bfruntime_lib.core_switch.CoreSwitch#get_telem_rtp_ip(udp_port) that obtains the IPv4 or 6 vaule for a given UDP value. If none, skip...

File: trans_sec/controller/ddos_sdn_controller.py [hunk 2]
```diff
@@ -45,14 +56,57 @@ def start(self, add_di):
...
+                                                 drop_tbl_keys=keys)
+                drop_pkt = drop_pkt / UDP(dport=consts.UDP_DRPT_DST_PORT, sport=consts.UDP_DRPT_SRC_PORT,
+                                          len=udp_len)
+                drop_pkt = drop_pkt / 'tps drop report'
+                try:
+                    sendp(drop_pkt, verbose=2)
+                    logger.info("Sent Drop Report packet")
+                except Exception as e:
+                    logger.info("Unable to send drop report - [%s]", e)
+        threading.Timer(10.0, self.create_drop_report).start()
```
Comment (spisarski) [inline trans_sec/controller/ddos_sdn_controller.py:108]: Can you instantiate the timer in the constructor and store it into a variable named something like:
self.drop_rpt_thread = threading.Timer(...)

Then in start line #59, replace with:
self.drop_rpt_thread.start()

File: trans_sec/controller/ddos_sdn_controller.py [hunk 3]
```diff
@@ -45,14 +56,57 @@ def start(self, add_di):
         logger.info('Starting HTTP server on port - [%s]',
                     self.http_server.port)
         self.http_server.start()
-
+        self.create_drop_report()
```
Comment (spisarski) [inline trans_sec/controller/ddos_sdn_controller.py:59]: also see comment on line #108.
Replace with
self.drop_rpt_thread.start()

File: trans_sec/controller/ddos_sdn_controller.py [hunk 4]
```diff
@@ -45,14 +60,68 @@ def start(self, add_di):
...
                     self.http_server.port)
         self.http_server.start()
-
+        self.drop_rpt_thread.start()
         self.__main_loop()
 
     def stop(self):
         logger.info('Stopping SDN controller')
         self.running = False
         self.http_server.stop()
```
Comment (spisarski) [inline trans_sec/controller/ddos_sdn_controller.py:69]: We should stop the thread too
self.drop_rpt_thread.stop()

File: trans_sec/controller/ddos_sdn_controller.py [hunk 5]
```diff
@@ -45,14 +60,68 @@ def start(self, add_di):
...
 
+    def create_drop_report(self):
+        while True:
+            self.send_drop_report()
+            sleep(10)
```
Comment (spisarski) [inline trans_sec/controller/ddos_sdn_controller.py:74]: You should make this interval an input parameter.

**Review by spisarski — APPROVED** [review]
Please address the two additional comments either on this PR and I won't need to re-review as long as CI completes else just go ahead and merge and create a ticket to address those two minor issues.

**Pull request merged**